quil-rs/src/instruction/extern_call.rs

```diff
@@ -0,0 +1,2060 @@
+            crate::parser::reserved_pragma_extern::parse_extern_signature(
+                signature_tokens.as_slice(),
+            )
+            .map_err(crate::parser::ParseError::from_nom_internal_err),
+        )
+        .map_err(ExternSignatureError::Syntax)?;
+        Ok(signature)
+    }
+}
+
+impl Quil for ExternSignature {
+    fn write(
+        &self,
+        writer: &mut impl std::fmt::Write,
+        fall_back_to_debug: bool,
+    ) -> Result<(), crate::quil::ToQuilError> {
+        if let Some(return_type) = &self.return_type {
+            return_type.write(writer, fall_back_to_debug)?;
+            if !self.parameters.is_empty() {
+                write!(writer, " ")?;
+            }
+        }
+        if self.parameters.is_empty() {
+            return Ok(());
+        }
+        write!(writer, "(")?;
+        for (i, parameter) in self.parameters.iter().enumerate() {
+            if i > 0 {
+                write!(writer, ", ")?;
+            }
+            parameter.write(writer, fall_back_to_debug)?;
+        }
+        write!(writer, ")").map_err(Into::into)
+    }
+}
+
+/// An extern definition with a name and optional signature. Note, this is not a
+/// Quil instruction or command, though it may become so in the future. Currently,
+/// it is defined as a reserved pragma.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct ExternDefinition {
+    /// The name of the extern definition. This must be a valid user identifier.
+    pub name: String,
+    /// The signature of the extern definition, if any.
+    pub signature: Option<ExternSignature>,
+}
+
+impl ExternDefinition {
+    pub fn try_new(
+        name: String,
+        signature: Option<ExternSignature>,
+    ) -> Result<Self, ExternValidationError> {
+        validate_user_identifier(name.as_str()).map_err(ExternValidationError::Name)?;
+
+        Ok(Self { name, signature })
+    }
+}
+
+impl Quil for ExternDefinition {
+    fn write(
+        &self,
+        writer: &mut impl std::fmt::Write,
+        fall_back_to_debug: bool,
+    ) -> Result<(), crate::quil::ToQuilError> {
+        write!(writer, "{}", self.name)?;
+        if let Some(signature) = &self.signature {
+            write!(writer, " \"")?;
+            signature.write(writer, fall_back_to_debug)?;
+            write!(writer, "\"")?;
+        }
+        Ok(())
+    }
+}
+
+/// An error that can occur when validating an extern definition.
+#[derive(Debug, thiserror::Error, PartialEq)]
+pub enum ExternValidationError {
+    /// The specified name is not a valid user identifier.
+    #[error(transparent)]
+    Name(#[from] IdentifierValidationError),
+    /// There are more than one extern definitions with the same name.
+    #[error("duplicate extern definition {0}")]
+    Duplicate(String),
+    /// The extern definition has a signature but it lacks a return or parameters.
+    #[error("extern definition {0} has a signature but it lacks a return or parameters")]
```

Comment:
What signature can it have, then?

Reply:
This is an odd part of the spec, which basically requires that the `CALL` has at least one argument. I get that the intent is to require that statefulness is maintained in the user memory space, however, this seems a futile requirement to try to enforce because a user could easily workaround it:

```
DECLARE useless_bit BIT[1]
PRAGMA EXTERN foo "(ignored : BIT[1])"
CALL foo useless_bit[0] # the bit is ignored and the function mutates state
```

I'm partial to dropping this requirement in the implementation as well as the spec.